### PR TITLE
Alerting: Use interfaces for the multi-org Alertmanager

### PR DESCRIPTION
**Pull request by santihernandezc**
This PR replaces concrete structs of the multi-org Alertmanager for interfaces describing the functionality that different parts of Grafana need from that struct.

This will allow us to replace the multi-org Alertmanager with a struct that acts as a proxy, routing requests to an external Alertmanager, ultimately disabling the internal one altogether.

### CODE REVIEW

File: pkg/services/ngalert/api/api_alertmanager.go [hunk 1]
```diff
@@ -27,10 +27,18 @@ const (
...
+	ActivateHistoricalConfiguration(ctx context.Context, orgId int64, id int64) error
+	ApplyAlertmanagerConfiguration(ctx context.Context, org int64, config apimodels.PostableUserConfig) error
+	AlertmanagerFor(orgID int64) (*notifier.Alertmanager, error)
+}
+
 type AlertmanagerSrv struct {
 	log    log.Logger
 	ac     accesscontrol.AccessControl
-	mam    *notifier.MultiOrgAlertmanager
+	moa    multiOrgAlertmanager
```
Comment (santihernandezc) [inline pkg/services/ngalert/api/api_alertmanager.go:41]: From here to the end of this file I'm just replacing `mam -> moa`

File: pkg/services/ngalert/sender/router_test.go [hunk 2]
```diff
@@ -61,7 +61,7 @@ func TestIntegrationSendingToExternalAlertmanager(t *testing.T) {
 			"implementation":             "prometheus",
 		}),
 	}
-	alertsRouter := NewAlertsRouter(moa, fakeAdminConfigStore, mockedClock, appUrl, map[int64]struct{}{}, 10*time.Minute,
+	alertsRouter := NewAlertsRouter(moa.AlertmanagerFor, fakeAdminConfigStore, mockedClock, appUrl, map[int64]struct{}{}, 10*time.Minute,
```
Comment (santihernandezc) [inline pkg/services/ngalert/sender/router_test.go:64]: The only change in this file is passing a function instead of the struct to `NewAlertsRouter()`.

**Comment by yuri-tceretian** [comment]
> This will allow us to replace the multi-org Alertmanager with a struct that acts as a proxy, routing requests to an external Alertmanager, ultimately disabling the internal one altogether.

Why not replace notifier.Alertmanager with cloud implementation?

**Comment by santihernandezc** [comment]
> > This will allow us to replace the multi-org Alertmanager with a struct that acts as a proxy, routing requests to an external Alertmanager, ultimately disabling the internal one altogether.
> 
> Why not replace notifier.Alertmanager with cloud implementation?

@yuri-tceretian wdym by "cloud implementation"? The plan is to use a struct that implements the same interfaces and acts as a proxy to the external Alertmanager, using a client to make HTTP requests.

**Comment by yuri-tceretian** [comment]
The multi-org Alertmanager contains logic that performs two tasks:
- maintain the configuration in the database
- configure alertmanager per organization and keep it in-sync with the database.
The latter function is the main one. 

In the cloud alertmaanger, there will not be any organization support within a single tenant. Therefore, in my opinion, it is more natural to have an abstraction of a single alertmanager and implement a client for the cloud alertmanager that will implement this abstraction, and keep multi-org alertmanager the same. 
The first function of the multiorg alertmanager can be useful too because we're working on a new API for managing contact points, notification policies, and other pieces of AM config separately instead of having UI submit an entire blob of data, and that layer could be useful.

**Comment by santihernandezc** [comment]
@yuri-tceretian 

> In the cloud alertmaanger, there will not be any organization support within a single tenant. Therefore, in my opinion, it is more natural to have an abstraction of a single alertmanager and implement a client for the cloud alertmanager that will implement this abstraction, and keep multi-org alertmanager the same.

This is also true for any Grafana Cloud customer. We're using the multi-tenant Alertmanager struct even though there's only one org per instance.

The "multi-org" Alertmanager is actually one single Alertmanager in Cloud. The intention is to keep things as close to what they currently are as possible.

> The first function of the multiorg alertmanager can be useful too...

When using the Cloud Alertmanager no configuration will be kept in the database. Every change will be made against the Cloud Alertmanager's API and saved in object storage. The only time we would write Alertmanager changes to our database is when and if we need to roll back and use the internal Alertmanager again.

**Comment by santihernandezc** [comment]
Closing in favor of https://github.com/grafana/grafana/pull/73900

**Pull request closed without merging**